COLLABORATOR_GUIDE.md

```diff
@@ -84,6 +87,197 @@ All pull requests that modify executable code should be subjected to
+  and is not documented as part of the Node.js Public API, is considered to be
+  an internal property.
+- Any object, property, method, argument, or event not documented in the
+  Node.js documentation is considered to be internal.
+- Any native C/C++ APIs/ABIs exported by the Node.js `*.h` header files that
+  are hidden behind the `NODE_WANT_INTERNALS` flag are considered to be
+  internal.
+
+Exception to each of these points can be made if use of a given internal API
+can be demonstrated to be sufficiently relied upon by the Node.js ecosystem
+such that any changes would cause too much breakage. The threshhold for what
+qualifies as "too much breakage" is to be decided on a case-by-case basis by
+the CTC.
+
+If it is determined that a currently undocumented object, property, method,
+argument, or event *should* be documented, then a pull request adding the
+documentation is required in order for it to be considered part of the "public"
+API.
+
+Breaking changes to internal elements are permitted in semver-patch or
+semver-minor commits but Collaborators should take significant care when
+making and reviewing such changes. Before landing such commits, an effort
+must be made to determine the potential impact of the change in the ecosystem
+by analyzing current use and by validating such changes through ecosystem
+testing using the [Canary in the Goldmine](https://github.com/nodejs/citgm)
+tool. If a change cannot be made without ecosystem breakage, then CTC review is
+required before landing the change as anything less than semver-major.
+
+If a determination is made that a particular internal API (for instance, an
+underscore `_` prefixed property) is sufficiently relied upon by the ecosystem
+such that any changes may break user code, then serious consideration should be
+given to providing an alternative Public API for that functionality before any
+breaking changes are made.
+
+### Breaking Changes
+
+Backwards-incompatible changes may land on the master branch at any time after
+sufficient review by collaborators and approval of at least two CTC members.
+
+Examples of breaking changes include, but are not necessarily limited to,
+removal or redefinition of existing API arguments, changing return values
+(except when return values do not currently exist), adding or removing errors,
+changing error messages in any way, altering expected timing of an event (e.g.
+moving from sync to async responses or vice versa), and changing the
+non-internal side effects of using a particular API.
+
+With a few notable exceptions outlined below, when backwards incompatible
+changes to a *Public* API are necessary, the existing API *must* be deprecated
+*first* and the new API either introduced in parallel or added after the next
+major Node.js version following the deprecation as a replacement for the
+deprecated API. In other words, as a general rule, existing *Public* APIs
```

Comment:
One challenge here is that, at present, there is no such contract between Node.js and the VMs that any VM provided API that Node.js exposes must go through the same 6mo+ deprecation cycle. Either we need to co-ordinate with VM on being able to offer this guarantee on exposed VM APIs, or we should explicitly call this out in the exceptions below.

In practice, as in the old debug protocol example recently, we have the ability to work/negotiate with the VMs to figure out the timeline – the problems is that the VMs may not necessarily know what parts of their APIs are being exposed by Node, and the kind of a deprecation guarantee Node intends to provide for these APIs that the VM needs to uphold.